File: frontend/src/lib/m365-operative.ts
/** Microsoft 365 Operativo — tipos API Fase 6 */

export interface M365SuggestedAction {
  key: string;
  label: string;
  description?: string;
  confidence?: number;
  href?: string;
  auto_eligible?: boolean;
}

export interface M365EmailRelation {
  entity_type: string;
  entity_id?: string | null;
  label: string;
  confidence?: number;
  href?: string | null;
  metadata?: Record<string, unknown>;
}

export interface M365ProcessedEmail {
  id: string;
  mailbox: string;
  external_message_id: string;
  subject: string;
  sender_email: string;
  sender_name?: string | null;
  received_at: string;
  body_preview?: string | null;
  classification: string;
  classification_label: string;
  classification_confidence: number;
  extracted_data: Record<string, unknown>;
  relations: M365EmailRelation[];
  suggested_actions: M365SuggestedAction[];
  attachments: Record<string, unknown>[];
  sharepoint_path?: string | null;
  processing_status: string;
  graph_connected: boolean;
  demo_source: boolean;
  hermes_indexed: boolean;
  related_dgcp_process_id?: string | null;
  related_task_id?: string | null;
  amount?: number | null;
  currency?: string | null;
}

export interface M365OperativeDashboard {
  graph_connected: boolean;
  demo_mode: boolean;
  mailboxes_monitored: number;
  emails_today: number;
  attachments_processed: number;
  quotes_detected: number;
  invoices_detected: number;
  purchase_orders_detected: number;
  dgcp_documents_detected: number;
  tasks_generated: number;
  documents_indexed: number;
  pending_actions: number;
  recent_emails: M365ProcessedEmail[];
  briefing_lines: string[];
  quick_actions: M365SuggestedAction[];
}

export interface M365ProcessedEmailListResponse {
  items: M365ProcessedEmail[];
  total: number;
  graph_connected: boolean;
  demo_mode: boolean;
  message?: string | null;
}

export interface M365SyncResponse {
  ingested: number;
  processed: number;
  skipped: number;
  graph_connected: boolean;
  demo_mode: boolean;
  message: string;
}

export interface M365ExecuteActionResponse {
  status: string;
  action_key: string;
  message: string;
  result: Record<string, unknown>;
  task_id?: string | null;
  notification_id?: string | null;
}

File: backend/app/schemas/m365_operative.py
"""Schemas — Microsoft 365 Operativo."""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field


class M365EmailAttachment(BaseModel):
    name: str
    content_type: str | None = None
    size_bytes: int | None = None
    extracted_text: str | None = None


class M365SuggestedAction(BaseModel):
    key: str
    label: str
    description: str | None = None
    confidence: int = 0
    href: str | None = None
    auto_eligible: bool = False


class M365EmailRelation(BaseModel):
    entity_type: str
    entity_id: str | None = None
    label: str
    confidence: int = 0
    href: str | None = None
    metadata: dict[str, Any] = Field(default_factory=dict)


class M365ProcessedEmailResponse(BaseModel):
    id: uuid.UUID
    mailbox: str
    external_message_id: str
    subject: str
    sender_email: str
    sender_name: str | None = None
    received_at: datetime
    body_preview: str | None = None
    classification: str
    classification_label: str
    classification_confidence: int
    extracted_data: dict[str, Any] = Field(default_factory=dict)
    relations: list[M365EmailRelation] = Field(default_factory=list)
    suggested_actions: list[M365SuggestedAction] = Field(default_factory=list)
    attachments: list[M365EmailAttachment] = Field(default_factory=list)
    sharepoint_path: str | None = None
    processing_status: str
    graph_connected: bool
    demo_source: bool
    hermes_indexed: bool
    related_dgcp_process_id: uuid.UUID | None = None
    related_task_id: uuid.UUID | None = None
    amount: Decimal | None = None
    currency: str | None = None


class M365ProcessedEmailListResponse(BaseModel):
    items: list[M365ProcessedEmailResponse]
    total: int
    graph_connected: bool
    demo_mode: bool
    message: str | None = None


class M365OperativeDashboard(BaseModel):
    graph_connected: bool
    demo_mode: bool
    mailboxes_monitored: int
    emails_today: int
    attachments_processed: int
    quotes_detected: int
    invoices_detected: int
    purchase_orders_detected: int
    dgcp_documents_detected: int
    tasks_generated: int
    documents_indexed: int
    pending_actions: int
    recent_emails: list[M365ProcessedEmailResponse] = Field(default_factory=list)
    briefing_lines: list[str] = Field(default_factory=list)
    quick_actions: list[M365SuggestedAction] = Field(default_factory=list)


class M365ExecuteActionRequest(BaseModel):
    action_key: str
    params: dict[str, Any] = Field(default_factory=dict)


class M365ExecuteActionResponse(BaseModel):
    status: str
    action_key: str
    message: str
    result: dict[str, Any] = Field(default_factory=dict)
    task_id: uuid.UUID | None = None
    notification_id: uuid.UUID | None = None


class M365InboundEmailWebhook(BaseModel):
    mailbox: str = "cotizaciones@justech.do"
    message_id: str | None = None
    subject: str
    sender_email: str
    sender_name: str | None = None
    body_text: str = ""
    received_at: datetime | None = None
    attachments: list[M365EmailAttachment] = Field(default_factory=list)


class M365CalendarSuggestion(BaseModel):
    title: str
    event_type: str
    suggested_date: datetime | None = None
    source_email_id: uuid.UUID | None = None
    related_entity_type: str | None = None
    related_entity_id: str | None = None
    confidence: int = 0


class M365CalendarSuggestionsResponse(BaseModel):
    items: list[M365CalendarSuggestion]
    total: int


class M365SyncResponse(BaseModel):
    ingested: int
    processed: int
    skipped: int
    graph_connected: bool
    demo_mode: bool
    message: str
